mne/preprocessing/tests/test_xdawn.py

```diff
@@ -0,0 +1,140 @@
+    # provide covariance object
+    signal_cov = compute_raw_data_covariance(raw, picks=picks)
+    xd = Xdawn(n_components=2, correct_overlap=False,
+               signal_cov=signal_cov, reg=None)
+    xd.fit(epochs)
+    # provide ndarray
+    signal_cov = np.eye(len(picks))
+    xd = Xdawn(n_components=2, correct_overlap=False,
+               signal_cov=signal_cov, reg=None)
+    xd.fit(epochs)
+    # provide ndarray of bad shape
+    signal_cov = np.eye(len(picks) - 1)
+    xd = Xdawn(n_components=2, correct_overlap=False,
+               signal_cov=signal_cov, reg=None)
+    assert_raises(ValueError, xd.fit, epochs)
+
+    # fit with baseline correction and ovverlapp correction should throw an
+    # error
+    epochs = Epochs(raw, events, event_id, tmin, tmax, picks=picks,
+                    preload=True, baseline=(None, 0), verbose=False)
+
+    xd = Xdawn(n_components=2, correct_overlap=True)
+    assert_raises(ValueError, xd.fit, epochs)
+
+
+def test_xdawn_apply_transform():
+    """Test Xdawn apply and transform."""
+    # get data
+    raw, events, picks = _get_data()
+    epochs = Epochs(raw, events, event_id, tmin, tmax, picks=picks,
+                    preload=True, baseline=None, verbose=False)
+    n_components = 2
+    # Fit Xdawn
+    xd = Xdawn(n_components=n_components, correct_overlap='auto')
+    xd.fit(epochs)
+
+    # apply on raw
+    xd.apply(raw)
+    # apply on epochs
+    xd.apply(epochs)
+    # apply on evoked
+    xd.apply(epochs.average())
+    # apply on other thing should raise an error
+    assert_raises(ValueError, xd.apply, 42)
+
+    # transform on epochs
+    xd.transform(epochs)
+    # transform on ndarray
+    xd.transform(epochs._data)
+    # transform on someting else
+    assert_raises(ValueError, xd.transform, 42)
+
+
+@requires_sklearn
+def test_xdawn_regularization():
+    """Test Xdawn with regularization."""
+    # get data
+    raw, events, picks = _get_data()
+    epochs = Epochs(raw, events, event_id, tmin, tmax, picks=picks,
+                    preload=True, baseline=None, verbose=False)
+
+    # test xdawn with overlap correction
+    xd = Xdawn(n_components=2, correct_overlap=True,
+               signal_cov=None, reg=0.1)
+    xd.fit(epochs)
+    # ========== with cov regularization ====================
+    # ledoit-wolf
+    xd = Xdawn(n_components=2, correct_overlap=False,
+               signal_cov=np.eye(len(picks)), reg='lws')
+    xd.fit(epochs)
+    # oas
+    xd = Xdawn(n_components=2, correct_overlap=False,
+               signal_cov=np.eye(len(picks)), reg='oas')
+    xd.fit(epochs)
+    # with shrinkage
+    xd = Xdawn(n_components=2, correct_overlap=False,
+               signal_cov=np.eye(len(picks)), reg=0.1)
+    xd.fit(epochs)
+    # with bad shrinkage
+    xd = Xdawn(n_components=2, correct_overlap=False,
+               signal_cov=np.eye(len(picks)), reg=2)
+    assert_raises(ValueError, xd.fit, epochs)
```

Comment:
do you want to add the directive `run_tests_if_main` at the bottom so that we can profile the timing and memory?